design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+<!-- /toc -->
+
+## Summary
+
+In order to implement policy and improve auditing introspection of certificates
+requested and signed by cert-manager, cert-manager must implement a better
+identity framework. Identity in this context is a user or machine which requests
+a signed certificate from cert-manager. This identity will be managed and
+enforced on CertificateRequest resources by cert-manager.
+
+
+## Motivation
+
+The identity of a requester forms a large part of an automated policy system
+that needs to decide whether a certificate should or shouldn't be signed. By
+creating a trusted source of identity for `CertificateRequest`s, a policy engine
+is able to determine whether that identity is allowed to request that
+certificate, given some policy configuration setup by a cluster administrator.
+
+Although auditing exists in Kubernetes and exposes the identity of the
+requester, its configuration is not always exposed to end users, such as
+when using managed Kubernetes (GKE, EKS etc). In scenarios where
+configuring auditing is available, it is often not preferable and an
+anti-pattern to make runtime decisions on historical audit logs.
+
+The upstream [Kubernetes certificates
+`CertificateSigningRequest`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#certificatesigningrequest-v1-certificates-k8s-io)
+API has identity as part of the resource. As the project intends to align and
+likely make this resource compatible with cert-manager in some way, the
+`CertificateRequest` should expose identical identity details. This means
+extensions or additions to cert-manager (policy, auditing) can have the same
+guarantees about what information is available from both resources.
+
+
+### Goals
+
+- Introduce identity to the `CertificateRequest` resource which is securely
+    enforced by cert-manager
+- Ensure the identity exposed in `CertificateRequest`s has parity with the
+    upstream Kubernetes `CertificateSigningRequest` resource
+- Enable external tooling or additions to cert-manager that can make use of this
+    identity in a useful way
+
+### Non-Goals
+
+- Discuss how this identity can be used (policy, auditing)
+- Make changes to upstream Kubernetes to implement identiy in cert-manager
+- Although considered below, "passed down" identity is not part of this design
+
+## Proposal
+
+Kubernetes does [not currently support immutable
+fields](https://github.com/kubernetes/enhancements/blob/8b9b994136371f1bc938aabf012f4c45535d684c/keps/sig-api-machinery/20190603-immutable-fields.md)
+for CRDs. The cert-manager webhook will be responsible for populating and
+enforcing identity fields which are present on the spec of `CertificateRequest`
+resources.
+
+The webhook will be responsible for enforcing the following during a CREATE
+operation:
+- No identity fields have been set by the user creating the resource
+- Set identity fields to exactly what is received from the API server in the
```

Comment:
Can we also link to the above area of code so we can be clear what's expected to be set? (https://github.com/kubernetes/kubernetes/blob/7a94debba5f8c21bbf8b42b2a7f1d5e974ddb837/pkg/registry/certificates/certificates/strategy.go#L63-L79)